Sources/NIO/EventLoopFuture.swift

```diff
@@ -957,6 +960,61 @@ extension EventLoopFuture {
     }
 }
 
+extension EventLoopFuture {
+    /// Returns a new `EventLoopFuture` that succeeds when all of the provided `EventLoopFuture`s complete.
+    /// The new `EventLoopFuture` will contain an array of results, maintaining ordering for each of the `EventLoopFuture`s.
+    ///
+    /// The returned `EventLoopFuture` always succeeds, regardless of any failures from the waiting futures.
+    ///
+    /// If it is desired to flatten them into a single `EventLoopFuture` that fails on the first `EventLoopFuture` failure,
+    /// use one of the `reduce` methods instead.
+    /// - Parameter futures: An array of homogenous `EventLoopFuture`s to gather results from.
+    /// - Returns: A new `EventLoopFuture` with all the results of the provided futures.
+    public static func whenAllComplete<InputValue>(_ futures: [EventLoopFuture<InputValue>],
+                                                   eventLoop: EventLoop) -> EventLoopFuture<[Result<InputValue, Error>]> {
+        let promise = eventLoop.makePromise(of: [Result<InputValue, Error>].self)
+
+        if eventLoop.inEventLoop {
+            _whenAllComplete0(promise, futures, eventLoop: eventLoop)
+        } else {
+            eventLoop.execute {
+                _whenAllComplete0(promise, futures, eventLoop: eventLoop)
+            }
+        }
+
+        return promise.futureResult
+    }
+
+    private static func _whenAllComplete0<InputValue>(_ promise: EventLoopPromise<[Result<InputValue, Error>]>,
+                                                      _ futures: [EventLoopFuture<InputValue>],
+                                                      eventLoop: EventLoop) {
+        eventLoop.assertInEventLoop()
+
+        var remainingCount = futures.count
+        var results: [Result<InputValue, Error>] = .init(repeating: .failure(OperationPlaceholderError()), count: futures.count)
+
+        // loop through the futures to chain callbacks to execute on the initiating event loop and grab their index
+        // in the "futures" to store their result in the same location
+        for (index, future) in futures.enumerated() {
+            future.hopTo(eventLoop: eventLoop)
+                .whenComplete { result in
+                    results[index] = result
+                    remainingCount -= 1
+
+                    guard remainingCount == 0 else { return }
+
+                    // verify that all operations have been completed
+                    assert(results.contains(where: {
+                        guard case let .failure(error) = $0 else { return false }
+                        return error is OperationPlaceholderError
+                    }) == false)
```

Comment:
Minor style nit, but I prefer to `assert(!` instead of `assert( == false)`. Explicit equality operations against booleans just read weirdly to me.

Reply:
Haha, I'm the other way, I find that `!results.contains(...)` to read weirdly - but I don't have to read this code as often as you :P